### ISSUE TITLE
Openvpn deferred authentication

### ISSUE THREAD

**Issue by csszep** [opnsense/core#3992]
It seems the same problem exist on Opnsense too...

https://redmine.pfsense.org/issues/7905

**Comment by csszep** [comment]
Hi!

OPnsense using the "auth-user-pass-verfiy" directive for user authentication which is blocking the main process especially when the remote authentication (radius) is slow .

There is a mechanism in openvpn plugin api called deferred authentication which solve this problem.

There is a "wontfix" ticket that explain the problem.

https://community.openvpn.net/openvpn/ticket/222

**Comment by camouflageX** [comment]
Hey, 

we are also having the problem with packet loss for all users when a new user authenticates.

It seems pfSense has this fixed since version 2.4.4 according to this: https://redmine.pfsense.org/issues/7905

It would be nice if you could also implement the new authentication plugin in OPNsense. Thank you very much!

**Comment by AdSchellevis** [comment]
This issue has been automatically timed-out (after 180 days of inactivity).

For more information about the policies for this repository,
please read https://github.com/opnsense/core/blob/master/CONTRIBUTING.md for further details.

If someone wants to step up and work on this issue,
just let us know, so we can reopen the issue and assign an owner to it.

**Comment by marcquark** [comment]
is this feature considered at all on the roadmap? it seems to me that in any environment with more than a handful of users, it could become problematic if every time somebody authenticates all traffic stalls for a few seconds (say because an LDAP or RADIUS server is slow to respond)

**Comment by fichtner** [comment]
Allegedly, this was resolved in OpenVPN 2.5 and we will migrate to this version on our way to next major release 21.1.

**Comment by marcquark** [comment]
Any way to help testing?

**Comment by csszep** [comment]
Hi!

I don’t see anything about this in the Openvpn 2.5 release notes. The auth-user-pass-verfiy directive still does not support deferred authentication. Only auth-pam plugin and client-connect scripts received the feature.
You still need to develop an asynrchonous script designed to be called via a OpenVPN plugin api .

Or i miss something?

**Comment by csszep** [comment]
There is a patch targeted fot Openvpn 2.6 for auth-user-pass-verify deferred:

https://www.mail-archive.com/openvpn-devel@lists.sourceforge.net/msg21122.html

**Comment by marcquark** [comment]
@fichtner could you confirm whether adding something like a "sleep 5" [here](https://github.com/opnsense/core/blob/master/src/etc/inc/plugins.inc.d/openvpn/ovpn_auth_verify#L10) would be a viable way to test/provoke the situation?

i'd be willing to add that to my lab setup, then perform the test with 2.4 and 2.5 and report results.

**Comment by AdSchellevis** [comment]
@marcquark that should do the trick, but is very easy to test (1 client constant pinging, the other trying to connect)

**Comment by marcquark** [comment]
So i finally got round to testing this.

- Added an artificial 5 second delay to ovpn_auth_verify as described above.
- Connected one client and pinged OPNsense, then connected another client.

While the auth script blocks, traffic stalls on the first client on the stock installation. Installed the OpenVPN 2.5.0 package and rebooted. Same behaviour. So it looks like OpenVPN 2.5.0 does not fix this particular issue.

**Comment by AdSchellevis** [comment]
@marcquark this probably needs more time to figure out then. the release notes suggested it targeted this behaviour, if the current flow is fixable that would be preferable before looking into other alternatives, since the auth flow is also responsible for proper setting up client specific overrides.

**Comment by marcquark** [comment]
@AdSchellevis it's definitely still blocking even with OpenVPN 2.5.0

It's actually an easy fix to use the plugin as it is already available in ports. I think it makes sense to add it, considering that OpenVPN 2.5.0 only just came out and does **not** fix the issue. It only addresses up and down scripts, not auth-user-pass-verify (which is why the PR @csszep linked exists i guess). The change is easily reverted whenever a fixed ovpn version is released. The plugin has been successfully used in pfSense for years without problems, i wouldn't expect it to cause any issues whatsoever.

Would you mind reopening the ticket and adjusting the labels? U can assign it to me

**Comment by AdSchellevis** [comment]
@marcquark I do have doubts about using the work-around, since our implementation differs from the one in pfSense. we flush cso files in a similar way as the radius auth plugin for openvpn does (which means a change in flow might break peoples setup). And there's also the question if it wouldn't be better to switch to pam (https://github.com/opnsense/core/issues/3266) and how to unravel the rest of the auth flow in that case. 

If it would have been possible to backport an upstream patch in the meantime that would be nice, but in OpenVPN's git repo I couldn't find a patch handling async for auth-user-pass-verify.

I'll open this ticket now, but no guarantees we can spend enough time to pull this through in our community hours.

**Comment by marcquark** [comment]
@AdSchellevis hmm i can't follow. The flow isn't changed all that much from my point of view. The `ovpn_auth_verify` shell script is only a small wrapper, the actual authentication is done in `auth-user.php`. While the shell script is slightly adjusted to write the authentication result into a file instead of using the return value directly, the PHP script that does the actual work is left untouched and called in exactly the same way. What am i missing? I'm looking at the call `openvpn_csc_conf_write` now, is that where you're concerned things may break? Maybe i can dig deeper and figure this out and provide some more information

What are your thoughts about an optional toggle to switch between deferred and blocking authentication? And as a general question, would such an optional feature be implementable as a plugin rather than making changes to core? I feel like in this particular instance there's too much legacy code involved for that...

**Comment by AdSchellevis** [comment]
@marcquark I'm not sure if cso files being created on login keep working like they do when using the alternative auth sequence with the plugin, we have to test that. If it doesn't radius responses won't be handled properly anymore and manual cso's will cause issues as well. 

At a first glance a simple test case would be to add a manual cso with a custom tunnel network and see if this still works in async mode. The current code indeed has quite some legacy in it and particularly the authentication flow needs some attention (the current one isn't perfect either, it needs access to the config.xml, which ideally would be looser coupled).

Making this pluggable probably isn't worth the effort to be honest, the "plug" part should be the auth framework in my opinion, which has some challenges in this particular case.

Long story short, I'm not saying you're missing anything, I'm just not sure I'm currently overseeing the complete picture needed to distinct a quick fix (with future issues) and a step into a better solution. By the way, do you know if pam authentication in openvpn has the same concurrency limitations?

**Comment by marcquark** [comment]
@AdSchellevis thx for the clarification. Understand your points of concern.

I have gathered two more points that look reassuring, but are of course not definitive:
- Having read through the plugin's source, i'm certain that it doesn't do anything extraordinary. It does exactly what is described in [this article (found via OpenVPN Ticket #222)](https://engineering.freeagent.com/2017/05/22/external-authentication-scripts-in-openvpn-the-right-way/). In a nutshell: Double-Fork and do the actual authentication there, return `OPENVPN_PLUGIN_FUNC_DEFERRED` in order to not stall the OpenVPN main process.
- As far as writing the cso file to disk is concerned, i've double-checked and if i'm not badly overlooking something, the flow is very similar to pfSense. They flush the file to disk in [openvpn.auth-user.php](https://github.com/pfsense/pfsense/blob/1045afc5e7e904da4739345b25826e2c65984e67/src/etc/inc/openvpn.auth-user.php#L160) which is called by [ovpn_auth_verify_async](https://github.com/pfsense/pfsense/blob/1045afc5e7e904da4739345b25826e2c65984e67/src/usr/local/sbin/ovpn_auth_verify_async#L69). That chain and the one used here in OPNsense look very similar to me (or dare i say, identical under the cso lens). As mentioned in the PR, it has proven to work fine over there.
- I've done your suggested test (still with an artifical 10s delay in the auth script) and everything comes up fine.


Upon further reading i found a couple of interesting hints in the OpenVPN Source Code aswell:
https://github.com/OpenVPN/openvpn/blob/af0ae823242812d8be5d354d1ba3e88835e319cc/include/openvpn-plugin.h.in#L83
> [If OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY returned OPENVPN_PLUGIN_FUNC_DEFERRED,
> we don't proceed until authentication is verified via auth_control_file]

https://github.com/OpenVPN/openvpn/blob/af0ae823242812d8be5d354d1ba3e88835e319cc/include/openvpn-plugin.h.in#L100
> [If OPENVPN_PLUGIN_AUTH_USER_PASS_VERIFY returned OPENVPN_PLUGIN_FUNC_DEFERRED,
> we expect that authentication is verified via auth_control_file within
> the number of seconds defined by the "hand-window" option.  Data channel traffic
> will continue to flow uninterrupted during this period.]

Those comments are the same in master and 2.4


While i'd be surprised if those comments weren't telling the truth, i'm still looking further for convincing facts. Diving in:
Plugin authentication is handled directly before script authentication [here](https://github.com/OpenVPN/openvpn/blob/af0ae823242812d8be5d354d1ba3e88835e319cc/src/openvpn/ssl_verify.c#L1388). Following that path, we see that the main difference is that deferred authentication will set `ks->auth_deferred = true;`
At this point bottom-up leads to a blurry picture, so let's look at what happens with `ks->auth_deferred` top-down. The function that is called from the main event loop is [tls_multi_process](https://github.com/OpenVPN/openvpn/blob/af0ae823242812d8be5d354d1ba3e88835e319cc/src/openvpn/ssl.c#L3170), which in turn calls [tls_authentication_status](https://github.com/OpenVPN/openvpn/blob/af0ae823242812d8be5d354d1ba3e88835e319cc/src/openvpn/ssl_verify.c#L948). Bingo! That function is where the auth control file is checked and processed. TLS Authentication will not finish before the auth result is available (or times out i guess). Consequently nothing else will be done until then.


Conclusion:
Deferred authentication stalls during establishment of the TLS session. Before deferred auth is complete, the main event loop will not advance to calling other functions. That should be enough proof that deferred auth will not break existing setups, as it will behave just like a synchronous script call, but not block the main process.

Bonus points:
In 2.4 the PAM plugin suffers from the same problems. Deferred authentication support was added in 2.5: https://github.com/OpenVPN/openvpn/commit/c83b197a72a6f909a4ddcded027469f0da5d4a24 . All it takes is adding `setenv deferred_auth_pam 1` to the server config. So i guess it would be an easy transition to replace the script with PAM once 2.5 is used.

**Comment by marcquark** [comment]
Oh, one last thing i forgot to add. The documentation of tls_authentication_status in 2.4 is lackluster, but it's available in master: https://github.com/OpenVPN/openvpn/blob/3ac8e5923a12390f68aa901e04ab3204e326d243/src/openvpn/ssl_verify.h#L84

They changed a few things around and now use an enum instead of individual booleans to differentiate between authenticated, deferred and so on, but the logic has remained the same.

**Comment by mimugmail** [comment]
Hi, 

can we please reopen this one? I have a setup with nearly 250 concurrent users, authenticating against LDAP / AD. Since October last year we're getting constantly more and more reports about killed RDP connections inside the tunnel while the tunnel itself stays up. Also had a support call with @AdSchellevis troubleshooting sysctl's, cause my first guess was broken fibre nic during update.

Today we rebooted the domain controller and during this reboot one of the users tried to login which lead to a freeze for all 250 other users. I know there are workarounds like using multiple DC's, splitting users to different VPN processes, but the whole situation sounds weird, sadly. :)

If there's a way to speed things up, you know my company is always willing to help via sponsoring etc. :)

**Comment by csszep** [comment]
Hi!

One note:
Openvpn 2.6 will be released soon.
It will suport deferred auth via standard --auth-user-pass-verify script without any custom plugin or moving to pam.

https://github.com/OpenVPN/openvpn/blob/master/Changes.rst

Deferred auth support for scripts
    The --auth-user-pass-verify script supports now deferred authentication.

**Comment by fichtner** [comment]
@csszep thanks, do you have any idea what "soon" is? weeks? months? just to be able to think ahead on this

**Comment by csszep** [comment]
Hi!


OpenVPN 2.6.0 -- Released 25 January 2023

https://community.openvpn.net/openvpn/wiki/Downloads

**Comment by fichtner** [comment]
lucky us, thanks for the heads-up ❤️
